pkg/pub_package_reader/lib/src/pubspec_content_override.dart

```diff
@@ -0,0 +1,170 @@
+  'meta_types_json',
+  'meta_types_json_generator',
+  'meta_types_redux',
+  'meta_types_redux_generator',
+  'mocktail',
+  'mocktail_image_network',
+  'polymer_interop',
+};
+
+/// Override pubspec.yaml content if needed:
+///
+/// - If the archive was created before 2022-01-01, we may need to update the
+///   version number, version constraint and SDK constraints, as `pub_semver`
+///   accepted separator characters other than `.`.
+///   https://github.com/dart-lang/pub_semver/pull/63
+String overridePubspecContentIfNeeded({
+  required String content,
+  required DateTime published,
+}) {
+  // quick checks to skip overrides
+  if (published.year >= 2022) {
+    return content;
+  }
+
+  try {
+    return _fixupBrokenVersionAndConstraints(content);
+  } on FormatException catch (e, st) {
+    _logger.info('Unable to parse pubspec.', e, st);
+  }
+  // fallback: don't override content
+  return content;
+}
+
+/// Fixes broken version and version constraints in [pubspecYaml] and returns a new YAML string.
+///
+/// Prior to [pub_semver#63] version numbers and version constraints were accepted with any
+/// character as separate, instead of requiring  `.` (dot). This occurred because the regular expression
+/// in `package:pub_semver` didn't escape dots as `\.`, but just contained `.` (dot, meaning any character).
+///
+/// Thus, package versions published prior to 2022-01-01, maybe contain _version numbers_,
+/// _version constraints_ and _SDK constraints_ that doesn't use `.` (dot) as separator. As removing
+/// these _package versions_ would break existing users, this function will rewrite the version numbers
+/// and version constraints to have a valid format equivalent to how older versions of `pub_semver`
+/// would have interpreted the version number / version constraints.
+///
+/// [pub_semver#63]: https://github.com/dart-lang/pub_semver/pull/63
+String _fixupBrokenVersionAndConstraints(String pubspecYaml) {
+  final root = yaml.loadYaml(pubspecYaml);
+  if (root is! Map) {
+    _logger.warning(
+        'Unable to parse YAML for package, in _fixupBrokenVersionAndConstraints!');
+    return pubspecYaml; // return original as no changes can be made
+  }
+
+  // sanity check on the name again, now with parsed yaml
+  final name = root['name'];
+  if (name is! String || !_packagesWithBadVersions.contains(name)) {
+    return pubspecYaml; // return original as no changes needed
+  }
+
+  final editor = YamlEditor(pubspecYaml);
+  var hasBeenUpdated = false;
+  for (final c in _detectVersionEditCandidates(root)) {
+    final updated = c.value.replaceAllMapped(_lenientRegExp, (match) {
+      final major = int.parse(match[1]!);
+      final minor = int.parse(match[2]!);
+      final patch = int.parse(match[3]!);
+      final pre = match[5];
+      final build = match[8];
+      return Version(major, minor, patch, pre: pre, build: build).toString();
+    });
+    if (updated != c.value) {
+      editor.update(c.path, updated);
```

Comment:
If we do a modification, why don't we take `updated` and make:
```dart
try {
  Version.parse(updated);
} on FormatException catch (e, st) {
  _log.shout('Failed to fix broken version in package:$name', e, st);
  return pubspecYaml;
}
```
And similar for `VersionConstraint`.

It would be useful to assert that we're not producing garbage... because that might be harder to debug somewhere down the line.